torch/distributed/_tensor/placement_types.py

```diff
@@ -148,22 +148,24 @@ def _shard_tensor(
         self, tensor: torch.Tensor, mesh: DeviceMesh, mesh_dim: int
     ) -> torch.Tensor:
         """
-        shard and scatter a tensor on a mesh dimension (use coordinate
-        0 on the mesh dimension as source of truth)
+        copy the relevant slice of the global tensor to the local device.
         """
         my_coordinate = mesh.get_coordinate()
-        num_chunks = mesh.size(dim=mesh_dim)
-
         if my_coordinate is None:
             # if rank is not part of mesh, we simply return an empty tensor
-            return tensor.new_empty(0, requires_grad=tensor.requires_grad)
+            return tensor.new_empty(
+                0, device=mesh.device_type, requires_grad=tensor.requires_grad
+            )
 
+        num_chunks = mesh.size(dim=mesh_dim)
         scatter_list, pad_sizes = self._split_tensor(
             tensor, num_chunks, with_padding=True, contiguous=True
         )
+        src_shard = scatter_list[my_coordinate[mesh_dim]]
+        output = torch.empty_like(src_shard, device=mesh.device_type)
 
-        output = torch.empty_like(scatter_list[my_coordinate[mesh_dim]])
-        mesh.scatter(output, scatter_list, mesh_dim=mesh_dim)
+        if not tensor.is_meta:
+            output.copy_(src_shard)
```

Comment:
I don't think this is the correct way to do sharding, as we need to preserve a global tensor semantic, this means to also preserve the global tensor content. Otherwise if we simply just copy the corresponding slice on each rank, it would results in some correctness issue.

i.e. by default if you do `torch.randn` on each rank, it would result in different data as a result of different seed/rng of the rank, so each rank would produce a big tensor that purely based on their random seed. Given this, we can't simply copy corresponding slice as that would break the contract about a "global tensor" distribute to different devices.

Reply:
I see.

> i.e. by default if you do torch.randn on each rank, it would result in different data as a result of different seed/rng of the rank, so each rank would produce a big tensor that purely based on their random seed.

I see. yes, the approach of requiring the caller to call `torch.manual_seed` is a bit fragile right? also, it may not work for the multi-threaded case. I can revert it. 

But I noticed that every rank copies the entire tensor to the device:

https://github.com/pytorch/pytorch/blob/76af22103b4b886ca5d124d1df2e6cf8f53065f3/torch/distributed/_tensor/api.py#LL406C1-L407C52

 what do you think of doing this only in the source rank?

Reply:
i've reverted the change to use scatter once again

Reply:
based on our discussion, we will let every rank copy the entire tensor to preserve spmd-style programming.